File: packages/javascript/src/__tests__/topic-subscriber.spec.ts
import {
  topicSubscriber,
  topicSubscriberWildcard,
  Subscriber,
  TopicMetadata,
} from '../subscribers/subscriber';
import { Subscriber as SubscriberInterface } from '../types';
import config from '../core/config';

describe('Topic Subscriber', () => {
  beforeEach(() => {
    config.configure({
      env: 'test',
      appName: 'test-app',
    });
  });

  describe('topicSubscriber decorator', () => {
    it('should create subscriber for single topic', () => {
      @topicSubscriber('notifications')
      class NotificationSubscriber {
        async handle(_message: Record<string, unknown>, _metadata: TopicMetadata): Promise<void> {
          // Process notification
        }
      }

      const subscriber = new NotificationSubscriber() as unknown as SubscriberInterface;
      expect(subscriber.subjects).toEqual(['test.test-app.notifications']);
    });

    it('should create subscriber for hierarchical topic', () => {
      @topicSubscriber('notifications.email')
      class EmailNotificationSubscriber {
        async handle(_message: Record<string, unknown>, _metadata: TopicMetadata): Promise<void> {
          // Process email notification
        }
      }

      const subscriber = new EmailNotificationSubscriber() as unknown as SubscriberInterface;
      expect(subscriber.subjects).toEqual(['test.test-app.notifications.email']);
    });

    it('should create subscriber for multiple topics', () => {
      @topicSubscriber(['notifications', 'audit.events'])
      class MultiTopicSubscriber {
        async handle(_message: Record<string, unknown>, _metadata: TopicMetadata): Promise<void> {
          // Process multiple topics
        }
      }

      const subscriber = new MultiTopicSubscriber() as unknown as SubscriberInterface;
      expect(subscriber.subjects).toEqual([
        'test.test-app.notifications',
        'test.test-app.audit.events',
      ]);
    });

    it('should handle topic with wildcard', () => {
      @topicSubscriber('notifications.*')
      class WildcardSubscriber {
        async handle(_message: Record<string, unknown>, _metadata: TopicMetadata): Promise<void> {
          // Process wildcard topics
        }
      }

      const subscriber = new WildcardSubscriber() as unknown as SubscriberInterface;
      expect(subscriber.subjects).toEqual(['test.test-app.notifications.*']);
    });

    it('should normalize topic names', () => {
      @topicSubscriber('USER@EVENTS!')
      class NormalizedSubscriber {
        async handle(_message: Record<string, unknown>, _metadata: TopicMetadata): Promise<void> {
          // Process normalized topic
        }
      }

      const subscriber = new NormalizedSubscriber() as unknown as SubscriberInterface;
      expect(subscriber.subjects).toEqual(['test.test-app.user_events_']);
    });

    it('should preserve dots in hierarchical topics', () => {
      @topicSubscriber('analytics.user.signup')
      class AnalyticsSubscriber {
        async handle(_message: Record<string, unknown>, _metadata: TopicMetadata): Promise<void> {
          // Process analytics event
        }
      }

      const subscriber = new AnalyticsSubscriber() as unknown as SubscriberInterface;
      expect(subscriber.subjects).toEqual(['test.test-app.analytics.user.signup']);
    });

    it('should pass through subscription options', () => {
      @topicSubscriber('notifications', { ackWait: 60000 })
      class OptionsSubscriber {
        async handle(_message: Record<string, unknown>, _metadata: TopicMetadata): Promise<void> {
          // Process with options
        }
      }

      const subscriber = new OptionsSubscriber() as unknown as SubscriberInterface;
      expect(subscriber.options).toEqual({ ackWait: 60000 });
    });
  });

  describe('topicSubscriberWildcard decorator', () => {
    it('should create subscriber with wildcard pattern', () => {
      @topicSubscriberWildcard('notifications')
      class AllNotificationsSubscriber {
        async handle(_message: Record<string, unknown>, _metadata: TopicMetadata): Promise<void> {
          // Process all notifications
        }
      }

      const subscriber = new AllNotificationsSubscriber() as unknown as SubscriberInterface;
      expect(subscriber.subjects).toEqual(['test.test-app.notifications.>']);
    });

    it('should create subscriber for nested wildcard', () => {
      @topicSubscriberWildcard('order.processing')
      class OrderProcessingSubscriber {
        async handle(_message: Record<string, unknown>, _metadata: TopicMetadata): Promise<void> {
          // Process order events
        }
      }

      const subscriber = new OrderProcessingSubscriber() as unknown as SubscriberInterface;
      expect(subscriber.subjects).toEqual(['test.test-app.order.processing.>']);
    });

    it('should pass through subscription options', () => {
      @topicSubscriberWildcard('notifications', { ackWait: 60000 })
      class WildcardOptionsSubscriber {
        async handle(_message: Record<string, unknown>, _metadata: TopicMetadata): Promise<void> {
          // Process with options
        }
      }

      const subscriber = new WildcardOptionsSubscriber() as unknown as SubscriberInterface;
      expect(subscriber.options).toEqual({ ackWait: 60000 });
    });
  });

  describe('Subscriber', () => {
    it('should initialize with subject', () => {
      class TestSubscriber extends Subscriber {
        constructor() {
          super('test.test-app.notifications');
        }

        async handle(_event: Record<string, unknown>, _metadata: TopicMetadata): Promise<void> {
          // Test implementation
        }
      }

      const subscriber = new TestSubscriber();
      expect(subscriber.subjects).toEqual(['test.test-app.notifications']);
    });

    it('should initialize with multiple subjects', () => {
      class TestSubscriber extends Subscriber {
        constructor() {
          super(['test.test-app.notifications', 'test.test-app.audit']);
        }

        async handle(_event: Record<string, unknown>, _metadata: TopicMetadata): Promise<void> {
          // Test implementation
        }
      }

      const subscriber = new TestSubscriber();
      expect(subscriber.subjects).toEqual(['test.test-app.notifications', 'test.test-app.audit']);
    });

    it('should extract topic from subject', () => {
      class TestSubscriber extends Subscriber {
        constructor() {
          super('test.test-app.notifications.email');
        }

        async handle(_event: Record<string, unknown>, _metadata: TopicMetadata): Promise<void> {
          // Test implementation
        }

        public testExtractTopic(subject: string): string {
          return this.extractTopic(subject);
        }
      }

      const subscriber = new TestSubscriber();
      expect(subscriber.testExtractTopic('test.test-app.notifications.email')).toBe(
        'notifications.email'
      );
      expect(subscriber.testExtractTopic('production.prod-app.order.processing.completed')).toBe(
        'order.processing.completed'
      );
    });

    it('should check if message is from specific topic', () => {
      class TestSubscriber extends Subscriber {
        constructor() {
          super('test.test-app.notifications');
        }

        async handle(_event: Record<string, unknown>, _metadata: TopicMetadata): Promise<void> {
          // Test implementation
        }

        public testFromTopic(metadata: TopicMetadata, topicName: string): boolean {
          return this.fromTopic(metadata, topicName);
        }
      }

      const subscriber = new TestSubscriber();
      const metadata: TopicMetadata = {
        topic: 'notifications.email',
        subject: 'test.test-app.notifications.email',
        event_id: '123',
        domain: 'notifications',
        resource: 'notification',
        action: 'email',
        deliveries: 1,
      };

      expect(subscriber.testFromTopic(metadata, 'notifications.email')).toBe(true);
      expect(subscriber.testFromTopic(metadata, 'notifications')).toBe(false);
    });
  });

  describe('cross-environment support', () => {
    it('should work with different environments', () => {
      config.configure({
        env: 'production',
        appName: 'prod-app',
      });

      @topicSubscriber('notifications')
      class ProdSubscriber {
        async handle(_message: Record<string, unknown>, _metadata: TopicMetadata): Promise<void> {
          // Process notification in production
        }
      }

      const subscriber = new ProdSubscriber() as unknown as SubscriberInterface;
      expect(subscriber.subjects).toEqual(['production.prod-app.notifications']);
    });

    it('should create wildcard subscribers in different environments', () => {
      config.configure({
        env: 'staging',
        appName: 'staging-app',
      });

      @topicSubscriberWildcard('notifications')
      class StagingSubscriber {
        async handle(_message: Record<string, unknown>, _metadata: TopicMetadata): Promise<void> {
          // Process notification in staging
        }
      }

      const subscriber = new StagingSubscriber() as unknown as SubscriberInterface;
      expect(subscriber.subjects).toEqual(['staging.staging-app.notifications.>']);
    });
  });
});

File: packages/ruby/spec/topic_subscriber_spec.rb
# frozen_string_literal: true

require 'nats_pubsub'

RSpec.describe NatsPubsub::Subscriber, 'topic functionality' do
  before do
    NatsPubsub.reset!
    NatsPubsub.configure do |c|
      c.env = 'test'
      c.app_name = 'test-app'
    end
  end

  after { NatsPubsub.reset! }

  describe 'subscribe_to_topic' do
    it 'subscribes to a simple topic' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topic 'notifications'

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.length).to eq(1)
      expect(subscriptions.first[:pattern]).to eq('test.test-app.notifications')
      expect(subscriptions.first[:topic]).to eq('notifications')
      expect(subscriptions.first[:type]).to eq(:topic)
    end

    it 'subscribes to a hierarchical topic' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topic 'notifications.email'

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.first[:pattern]).to eq('test.test-app.notifications.email')
      expect(subscriptions.first[:topic]).to eq('notifications.email')
    end

    it 'subscribes to multi-level hierarchical topic' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topic 'order.processing.completed'

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.first[:pattern]).to eq('test.test-app.order.processing.completed')
    end

    it 'normalizes topic names' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topic 'USER@EVENTS!'

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.first[:pattern]).to eq('test.test-app.user_events_')
    end

    it 'preserves dots in hierarchical topics' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topic 'analytics.user.signup'

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.first[:pattern]).to eq('test.test-app.analytics.user.signup')
    end

    it 'preserves wildcards in topic names' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topic 'notifications.*'

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.first[:pattern]).to eq('test.test-app.notifications.*')
    end

    it 'accepts subscription options' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topic 'notifications', ack_wait: 60

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.first[:options]).to eq({ ack_wait: 60 })
    end
  end

  describe 'subscribe_to_topics' do
    it 'subscribes to multiple topics' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topics 'notifications', 'audit.events'

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.length).to eq(2)
      expect(subscriptions[0][:pattern]).to eq('test.test-app.notifications')
      expect(subscriptions[1][:pattern]).to eq('test.test-app.audit.events')
    end

    it 'accepts subscription options' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topics 'topic1', 'topic2', ack_wait: 60

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.length).to eq(2)
      expect(subscriptions[0][:options]).to eq({ ack_wait: 60 })
      expect(subscriptions[1][:options]).to eq({ ack_wait: 60 })
    end
  end

  describe 'subscribe_to_topic_wildcard' do
    it 'subscribes with wildcard pattern' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topic_wildcard 'notifications'

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.length).to eq(1)
      expect(subscriptions.first[:pattern]).to eq('test.test-app.notifications.>')
      expect(subscriptions.first[:topic]).to eq('notifications')
      expect(subscriptions.first[:type]).to eq(:topic_wildcard)
    end

    it 'subscribes to nested wildcard' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topic_wildcard 'order.processing'

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.first[:pattern]).to eq('test.test-app.order.processing.>')
    end

    it 'accepts subscription options' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topic_wildcard 'notifications', ack_wait: 60

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.first[:options]).to eq({ ack_wait: 60 })
    end
  end

  describe 'subscribe_to_event' do
    it 'maps domain/resource/action to topic' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_event domain: 'users', resource: 'user', action: 'created'

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.length).to eq(1)
      expect(subscriptions.first[:pattern]).to eq('test.test-app.users.user.created')
      expect(subscriptions.first[:topic]).to eq('users.user.created')
    end

    it 'supports wildcard actions' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_event domain: 'users', resource: 'user', action: '*'

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.first[:pattern]).to eq('test.test-app.users.user.*')
    end

    it 'accepts subscription options' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_event domain: 'users', resource: 'user', action: 'created', ack_wait: 60

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.first[:options]).to eq({ ack_wait: 60 })
    end
  end

  describe 'jetstream_options' do
    it 'sets default jetstream options' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topic 'notifications'

        def call(message, metadata); end
      end

      options = subscriber_class.jetstream_options
      expect(options).to include(
        retry: 5,
        ack_wait: 30,
        max_deliver: 5,
        dead_letter: true,
        batch_size: 25
      )
    end

    it 'allows customization of jetstream options' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topic 'notifications'
        jetstream_options retry: 3, ack_wait: 60

        def call(message, metadata); end
      end

      options = subscriber_class.jetstream_options
      expect(options).to include(
        retry: 3,
        ack_wait: 60
      )
    end
  end

  describe 'all_subscriptions' do
    it 'combines topic and legacy subscriptions' do
      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topic 'notifications'
        subscribe_to 'test.myapp.users.user.*'

        def call(message, metadata); end
      end

      all = subscriber_class.all_subscriptions
      expect(all.length).to eq(2)
      expect(all[0][:pattern]).to eq('test.test-app.notifications')
      expect(all[1][:pattern]).to eq('test.myapp.users.user.*')
    end
  end

  describe 'instance methods' do
    let(:subscriber_class) do
      Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topics 'notifications.email', 'notifications.sms'

        def call(message, metadata)
          # Custom implementation
        end
      end
    end

    let(:subscriber) { subscriber_class.new }

    describe '#from_topic?' do
      it 'returns true for matching topic' do
        metadata = { topic: 'notifications.email' }
        expect(subscriber.from_topic?(metadata, 'notifications.email')).to be true
      end

      it 'returns false for non-matching topic' do
        metadata = { topic: 'notifications.email' }
        expect(subscriber.from_topic?(metadata, 'notifications.sms')).to be false
      end
    end

    describe '#from_event?' do
      it 'returns true for matching domain/resource/action' do
        metadata = {
          domain: 'users',
          resource: 'user',
          action: 'created'
        }
        expect(
          subscriber.from_event?(metadata, domain: 'users', resource: 'user', action: 'created')
        ).to be true
      end

      it 'returns false for non-matching event' do
        metadata = {
          domain: 'users',
          resource: 'user',
          action: 'created'
        }
        expect(
          subscriber.from_event?(metadata, domain: 'users', resource: 'user', action: 'updated')
        ).to be false
      end
    end

    describe '#logger' do
      it 'returns configured logger' do
        expect(subscriber.logger).not_to be_nil
      end
    end

    describe '#call' do
      it 'raises NotImplementedError if not overridden' do
        base_subscriber = Class.new do
          include NatsPubsub::Subscriber

          subscribe_to_topic 'test'
        end

        expect do
          base_subscriber.new.call({}, {})
        end.to raise_error(NotImplementedError)
      end
    end
  end

  describe 'cross-environment support' do
    it 'works with different environments' do
      NatsPubsub.configure do |c|
        c.env = 'production'
        c.app_name = 'prod-app'
      end

      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topic 'notifications'

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.first[:pattern]).to eq('production.prod-app.notifications')
    end

    it 'creates wildcard subscribers in different environments' do
      NatsPubsub.configure do |c|
        c.env = 'staging'
        c.app_name = 'staging-app'
      end

      subscriber_class = Class.new do
        include NatsPubsub::Subscriber

        subscribe_to_topic_wildcard 'notifications'

        def call(message, metadata); end
      end

      subscriptions = subscriber_class.topic_subscriptions
      expect(subscriptions.first[:pattern]).to eq('staging.staging-app.notifications.>')
    end
  end
end
